### ISSUE TITLE
Opaque ports not working in particular circumstances (e.g., with redis)

### ISSUE THREAD

**Issue by fullykubed** [linkerd/linkerd2#12382]
### What is the issue?

When using the linkerd proxy sidecars with redis, connections cannot be established between proxied pods and redis. 

This impacts both the main redis port 6379 and the sentinel port 26379. While I have only been testing with redis, I can imagine that this might impact other TCP based systems that try to use the opaque proxy method.

### How can it be reproduced?

1. Run an EKS cluster
2. Deploy the Cilium CNI `v1.15.2` (follow the instructions for Linkerd compat)
3. Deploy Linkerd via the Helm Chart
4. Run the [bitnami/redis](https://artifacthub.io/packages/helm/bitnami/redis) helm chart
5. Deploy with `sentinel.enable` = `true`
5. Deploy with the pod annotation `"config.linkerd.io/opaque-ports" = "6379,26379"` 
5. Ensure that linkerd sidecars are injected into the pods
5. Attempt to connect to redis from another meshed pod in the cluster, either directly or via a service

### Logs, error output, etc

[Gist with manifests from the cluster](https://gist.github.com/fullykubed/f7d1e189fe6fca94b0a36ebbdbae32c4)

Originally, I tried with the native sidecar enabled (`"config.alpha.linkerd.io/proxy-enable-native-sidecar" = "true"`). But I also tried with that off. Additionally, I tried explicitly listing the ports as opaque via ``"config.linkerd.io/opaque-ports" = "6379,26379"` . None of this allowed the ability to connect.

Connecting directly to the redis pods or any of the redis services (headless or otherwise) from a pod without the proxy sidecar did work successfully, so this appears to be linkerd-specific.

Note that the only thing changed between tests are the linkerd annotations. 

Interestingly, there were two different types of error logs depending on whether the native sidecar was used.

### With Native Sidecar

The destination controller produced the following errors repeatedly:

```text
linkerd-destination-665558cc86-dtdvd destination {"addr":":8086","component":"endpoint-profile-translator","context-ns":"authentik","context-pod":"redis-1f0f-node-1","level":"error","msg":"Failed to create endpoint for 10.0.109.108:6379: failed to read inbound port: failed to find LINKERD2_PROXY_INBOUND_LISTEN_ADDR environment variable in any container for given pod spec","remote":{"IP":"10.0.178.50","Port":57678,"Zone":""},"time":"2024-04-03T17:42:48Z"}
linkerd-destination-665558cc86-h5ndf destination {"addr":":8086","component":"endpoint-profile-translator","context-ns":"authentik","context-pod":"redis-1f0f-node-1","level":"error","msg":"Failed to create endpoint for 10.0.109.108:26379: failed to read inbound port: failed to find LINKERD2_PROXY_INBOUND_LISTEN_ADDR environment variable in any container for given pod spec","remote":{"IP":"10.0.236.153","Port":53612,"Zone":""},"time":"2024-04-03T17:42:48Z"}
```

I believe this was the cause for these errors in the proxy sidecars:

```text
redis-1f0f-node-1 linkerd-proxy {"timestamp":"[   344.709554s]","level":"WARN","fields":{"message":"Unexpected policy controller response; retrying with a backoff","grpc.status":"Deadline expired before operation could complete","grpc.message":"initial item not received within timeout"},"target":"linkerd_app::dst","threadId":"ThreadId(1)"}
redis-1f0f-node-0 linkerd-proxy {"timestamp":"[   798.626827s]","level":"WARN","fields":{"message":"Unexpected policy controller response; retrying with a backoff","grpc.status":"Deadline expired before operation could complete","grpc.message":"initial item not received within timeout"},"target":"linkerd_app::dst","threadId":"ThreadId(1)"}
```

This is especially odd as you can see on the manifest that the redis pods contain the sidecar with `LINKERD2_PROXY_INBOUND_LISTEN_ADDR` set.

### Without Native Sidecar

In this scenario, proxy sidecars for pods attempting to connect to redis via a service generated these logs:

```txt
authentik-server-75844c6565-zxnl7 linkerd-proxy {"timestamp":"[  3042.733362s]","level":"INFO","fields":{"message":"Connection closed","error":"logical service redis-1f0f-master.authentik.svc.cluster.local:6379: concrete service redis-1f0f-master.authentik.svc.cluster.local:6379: connect timed out after 1s","client.addr":"10.0.142.178:60802","server.addr":"172.20.165.119:6379"},"target":"linkerd_app_core::serve","spans":[{"name":"outbound"}],"threadId":"ThreadId(1)"}
```

Proxy sidecars for pods attempting to connect to pod IPs directly generated these logs:

```txt
redis-1f0f-node-1 linkerd-proxy {"timestamp":"[   135.096226s]","level":"INFO","fields":{"message":"Connection closed","error":"connect timed out after 1s","client.addr":"10.0.142.181:54794","server.addr":"10.0.109.106:26379"},"target":"linkerd_app_core::serve","spans":[{"name":"outbound"}],"threadId":"ThreadId(1)"}
```

### output of `linkerd check -o short`

```text
linkerd-version
---------------
‼ cli is up-to-date
    is running version 24.3.3 but the latest edge version is 24.3.5
    see https://linkerd.io/2/checks/#l5d-version-cli for hints

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 24.3.3 but the latest edge version is 24.3.5
    see https://linkerd.io/2/checks/#l5d-version-control for hints

linkerd-control-plane-proxy
---------------------------
‼ control plane proxies are up-to-date
    some proxies are not running the current version:
	* linkerd-destination-6857b7765d-hqnsl (edge-24.3.3)
	* linkerd-destination-6857b7765d-sd44j (edge-24.3.3)
	* linkerd-identity-6bd8448888-q4lgl (edge-24.3.3)
	* linkerd-identity-6bd8448888-v65sj (edge-24.3.3)
	* linkerd-proxy-injector-5c46fb7795-2pfp7 (edge-24.3.3)
	* linkerd-proxy-injector-5c46fb7795-b85qb (edge-24.3.3)
    see https://linkerd.io/2/checks/#l5d-cp-proxy-version for hints

linkerd-viz
-----------
‼ viz extension proxies are up-to-date
    some proxies are not running the current version:
	* metrics-api-5dcddd489c-ltn8f (edge-24.3.3)
	* prometheus-54747f959-r7sc8 (edge-24.3.3)
	* tap-7dfff8cbd4-v6gb4 (edge-24.3.3)
	* tap-injector-c764b7b87-2b8ml (edge-24.3.3)
	* web-59b74d6d5-j8krg (edge-24.3.3)
    see https://linkerd.io/2/checks/#l5d-viz-proxy-cp-version for hints

Status check results are √
```

### Environment

- Kubernetes version: `v1.29.1-eks-508b6b3`
- Cilium Version: `v1.15.2`
- Linkerd Version: `edge-24.3.3`
- Redis Chart Version: `19.0.2`
- Redis Version: `7.2.4`

### Possible solution

Adding the pod annotation `"config.linkerd.io/skip-inbound-ports" = "6379,26379"` resolved the issue, but obviously this is not ideal as this is not much better than disabling the proxy for the redis pods altogether.

### Additional context

I am happy to try any other proposed solution.

### Would you like to work on fixing this bug?

None

**Comment by fullykubed** [comment]
For additional context, the Cilium is running in the kube-proxy replacement mode and is the only CNI running on the nodes.

**Comment by Sierra1011** [comment]
We're also seeing this with essentially the same config as above, but on version 2024.4.4

**Comment by adleong** [comment]
Hi @fullykubed,

I was pretty curious about the error you were seeing when native sidecars were disabled so I dug in to reproduce it.  What I think is happening is that the Redis helm chart ships with a NetworkPolicy resource which restricts ingress into the redis pods to only certain ports (6379,26379) and Cilium enforces this NetworkPolicy.  When Linkerd is configured to use opaque ports, it sends all of its traffic directly to Linkerd's inbound port (4143) and then the inbound proxy forwards to the appropriate application port.

The problem is that the NetworkPolicy does not allow connections to port 4143 and so Cilium blocks the connection.  In general, using port based NetworkPolicy is somewhat incompatible with Linkerd's opaque ports feature since all opaque traffic will flow through the same port.

I would recommend either removing this NetworkPolicy entirely or updating it to allow traffic to port 4143.

**Comment by fullykubed** [comment]
@adleong You are a hero. The NetworkPolicy installed by the chart was absolutely the issue. I had not expected the chart to install one automatically so never even thought to look. Replacing it with one that allows traffic to flow through Linkerd resolved the issue.

@cablespaghetti I also verified that native sidecars with opaque ports appears to cause issues. However, I tested #12453 in the latest edge release and that appears to resolve the problem.

Closing this issue as it is completely resolved.
